chips/esp32-c3/src/timg.rs

```diff
@@ -0,0 +1,245 @@
+//! TimG Group driver.
+
+use kernel::hil::time::{self, Alarm, Counter, Ticks, Ticks64, Time};
+use kernel::utilities::cells::OptionalCell;
+use kernel::utilities::registers::interfaces::{ReadWriteable, Readable, Writeable};
+use kernel::utilities::registers::register_bitfields;
+use kernel::utilities::registers::{register_structs, ReadWrite};
+use kernel::utilities::StaticRef;
+use kernel::ErrorCode;
+
+pub const TIMG0_BASE: StaticRef<TimgRegisters> =
+    unsafe { StaticRef::new(0x6001_F000 as *const TimgRegisters) };
+
+pub const TIMG1_BASE: StaticRef<TimgRegisters> =
+    unsafe { StaticRef::new(0x6002_0000 as *const TimgRegisters) };
+
+/// 20MHz `Frequency`
+#[derive(Debug)]
+pub struct Freq20MHz;
+impl time::Frequency for Freq20MHz {
+    fn frequency() -> u32 {
+        20_000_000
+    }
+}
+
+register_structs! {
+    pub TimgRegisters {
+        (0x000 => t0config: ReadWrite<u32, CONFIG::Register>),
+        (0x004 => t0lo: ReadWrite<u32>),
+        (0x008 => t0hi: ReadWrite<u32>),
+        (0x00C => t0update: ReadWrite<u32>),
+        (0x010 => t0alarmlo: ReadWrite<u32>),
+        (0x014 => t0alarmhi: ReadWrite<u32>),
+        (0x018 => t0loadlo: ReadWrite<u32>),
+        (0x01C => t0loadhi: ReadWrite<u32>),
+        (0x020 => t0load: ReadWrite<u32>),
+
+        (0x024 => _reserved0),
+
+        (0x048 => wdtconfig0: ReadWrite<u32, WDTCONFIG0::Register>),
+        (0x04C => wdtconfig1: ReadWrite<u32, WDTCONFIG1::Register>),
+        (0x050 => wdtconfig2: ReadWrite<u32>),
+        (0x054 => wdtconfig3: ReadWrite<u32>),
+        (0x058 => wdtconfig4: ReadWrite<u32>),
+        (0x05C => wdtconfig5: ReadWrite<u32>),
+        (0x060 => wdtfeed: ReadWrite<u32>),
+        (0x064 => wdtwprotect: ReadWrite<u32>),
+
+        (0x068 => rtccalicfg: ReadWrite<u32, RTCCALICFG::Register>),
+        (0x06C => rtccalicfg1: ReadWrite<u32, RTCCALICFG1::Register>),
+
+        (0x070 => int_ena: ReadWrite<u32, INT::Register>),
+        (0x074 => int_raw: ReadWrite<u32, INT::Register>),
+        (0x078 => int_st: ReadWrite<u32, INT::Register>),
+        (0x07C => int_clr: ReadWrite<u32, INT::Register>),
+        (0x080 => @END),
```

Comment:
It seems that the address of the interrupts is the only difference. Maybe instead it's better to just make the current implementation dynamic instead of adding all this new code

Reply:
I'm not sure what you mean here.

Reply:
From what I can tell the only difference between the current implementation in `chips/esp32` and this one is that the `int_*` register addresses are different.

So instead of copying the entire file to just change some addresses, why not have a option to the `new()` function that specifies which address to use?

Something like:

```rust
fn new(... c3: bool) {
    ...
    self.c3.set(c3);
    ...
}
fn my_function() {
    if self.c3.get() {
        self.registers.int_en_c3.write(...);
    } else {
        self.registers.int_en.write(...);
    }
}
```

Does that make sense?

Reply:
Some registers from the ESP32 seem to overlap with the ones in C3. I have to check that again.